bulkclient/src/main/java/au/csiro/pathling/export/ws/BulkExportTemplate.java

```diff
@@ -0,0 +1,294 @@
+   */
+  @Value
+  class PoolingState implements State {
+
+    @Nonnull
+    URI poolingURI;
+
+    int transientErrorCount;
+
+    @Nonnull
+    Duration retryAfter;
+
+    @Nonnull
+    @Override
+    public State handle(@Nonnull final BulkExportAsyncService service) throws IOException {
+      try {
+        final AsyncResponse asyncResponse = service.checkStatus(poolingURI);
+        // reset transient error counter
+        if (asyncResponse instanceof BulkExportResponse) {
+          return handleFinalResponse((BulkExportResponse) asyncResponse);
+        } else {
+          return handleContinueResponse((AcceptedAsyncResponse) asyncResponse);
+        }
+      } catch (final HttpError ex) {
+        if (ex.isTransient()) {
+          // log retrying a transient error
+          return handleTransientError(ex);
+        } else if (WebUtils.HTTP_TOO_MANY_REQUESTS == ex.getStatusCode()) {
+          // handle too many requests
+          return handleTooManyRequests(ex);
+        } else {
+          log.debug("Pooling: Http error: {}", ex.getMessage());
+          throw ex;
+        }
+      }
+    }
+
+    @Nonnull
+    private PoolingState handleTransientError(@Nonnull final HttpError ex) {
+      if (transientErrorCount < config.getMaxTransientErrors()) {
+        log.debug("Pooling: Retrying transient error {} ouf of {} : '{}'",
+            transientErrorCount, config.getMaxTransientErrors(), ex.getMessage());
+        final Duration timeToSleep = computeTimeToSleep(ex.getRetryAfter(),
+            config.getTransientErrorDelay());
+        log.debug("Pooling: Sleeping for {} ms", timeToSleep.toMillis());
+        return new PoolingState(poolingURI, transientErrorCount + 1, timeToSleep);
+      } else {
+        log.debug("Pooling: giving up on retrying of transient error: {}", ex.getMessage());
+        throw ex;
+      }
+    }
+
+    @Nonnull
+    private PoolingState handleTooManyRequests(@Nonnull final HttpError ex) {
+      log.debug("Pooling: Got too many requests error with retry-after: '{}'",
+          ex.getRetryAfter().map(RetryValue::toString).orElse("na"));
+      final Duration timeToSleep = computeTimeToSleep(ex.getRetryAfter(),
+          config.getTooManyRequestsDelay());
+      log.debug("Pooling: Sleeping for {} ms", timeToSleep.toMillis());
+      return new PoolingState(poolingURI, transientErrorCount, timeToSleep);
+    }
+
+    @Nonnull
+    private PoolingState handleContinueResponse(
+        @Nonnull final AcceptedAsyncResponse acceptedResponse) {
+      log.debug("Pooling: progress: '{}', retry-after: {}",
+          acceptedResponse.getProgress().orElse("na"),
+          acceptedResponse.getRetryAfter().map(RetryValue::toString).orElse("na"));
+      final Duration timeToSleep = computeTimeToSleep(acceptedResponse.getRetryAfter(),
+          config.getMinPoolingDelay());
+      log.debug("Pooling: Sleeping for {} ms", timeToSleep.toMillis());
```

Comment:
Yes.
AsyncConfig:
  /**
   * The maxium delay between two consecutive status requests. This may override the value returned
   * from the server in the 'retry-after' header.
   */
  @Builder.Default
  Duration maxPoolingDelay = Duration.ofSeconds(60);

Reply:
Great, thanks.